src/protocols/secure_channel/PASESession.cpp

```diff
@@ -578,115 +651,131 @@ CHIP_ERROR PASESession::HandleMsg1_and_SendMsg2(const System::PacketBufferHandle
+    peer_verifier_len = tlvReader.GetLength();
+    SuccessOrExit(err = tlvReader.GetDataPtr(peer_verifier));
+
+    SuccessOrExit(err = mSpake2p.ComputeRoundTwo(Y, Y_len, verifier, &verifier_len));
+
+    err = mSpake2p.KeyConfirm(peer_verifier, peer_verifier_len);
+    if (err != CHIP_NO_ERROR)
+    {
+        spake2pErr = kProtocolCodeNoSharedRoot;
         SuccessOrExit(err);
     }
 
+    SuccessOrExit(err = mSpake2p.GetKeys(mKe, &mKeLen));
+
     {
-        Encoding::PacketBufferWriter bbuf(System::PacketBufferHandle::New(verifier_len));
-        VerifyOrExit(!bbuf.IsNull(), err = CHIP_ERROR_NO_MEMORY);
+        size_t data_len = EstimateTLVStructOverhead(verifier_len, 1);
+
+        System::PacketBufferHandle msg3 = System::PacketBufferHandle::New(data_len);
+        VerifyOrExit(!msg3.IsNull(), err = CHIP_ERROR_NO_MEMORY);
+
+        System::PacketBufferTLVWriter tlvWriter;
+        tlvWriter.Init(std::move(msg3));
 
-        bbuf.Put(verifier, verifier_len);
-        VerifyOrExit(bbuf.Fit(), err = CHIP_ERROR_NO_MEMORY);
+        TLV::TLVType outerContainerType = TLV::kTLVType_NotSpecified;
+        SuccessOrExit(err = tlvWriter.StartContainer(TLV::AnonymousTag, TLV::kTLVType_Structure, outerContainerType));
+        SuccessOrExit(err = tlvWriter.Put(TLV::ContextTag(1), ByteSpan(verifier, verifier_len)));
+        SuccessOrExit(err = tlvWriter.EndContainer(outerContainerType));
+        SuccessOrExit(err = tlvWriter.Finalize(&msg3));
+
+        mNextExpectedMsg = MsgType::StatusReport;
 
         // Call delegate to send the Msg3 to peer
-        err = mExchangeCtxt->SendMessage(Protocols::SecureChannel::MsgType::PASE_Spake2p3, bbuf.Finalize());
+        err = mExchangeCtxt->SendMessage(MsgType::PASE_Spake2p3, std::move(msg3), SendFlags(SendMessageFlags::kExpectResponse));
         SuccessOrExit(err);
     }
 
     ChipLogDetail(SecureChannel, "Sent spake2p msg3");
 
-    mPairingComplete = true;
-
-    // Forget our exchange, as no additional messages are expected from the peer
-    mExchangeCtxt = nullptr;
-
-    // Call delegate to indicate pairing completion
-    mDelegate->OnSessionEstablished();
-
 exit:
 
     if (err != CHIP_NO_ERROR)
     {
-        SendErrorMsg(spake2pErr);
+        SendStatusReport(spake2pErr);
     }
     return err;
 }
 
-CHIP_ERROR PASESession::HandleMsg3(const System::PacketBufferHandle & msg)
+CHIP_ERROR PASESession::HandleMsg3(System::PacketBufferHandle && msg)
 {
-    CHIP_ERROR err              = CHIP_NO_ERROR;
-    const uint8_t * hash        = msg->Start();
-    Spake2pErrorType spake2pErr = Spake2pErrorType::kUnexpected;
+    CHIP_ERROR err      = CHIP_NO_ERROR;
+    uint16_t spake2pErr = kProtocolCodeInvalidParam;
 
     ChipLogDetail(SecureChannel, "Received spake2p msg3");
 
-    // We will set NextExpectedMsg to PASE_Spake2pError in all cases
+    // We will set NextExpectedMsg to StatusReport in all cases
```

Comment:
Why `set NextExpectedMsg to StatusReport in all cases` ?

After receiving Pake3, you should receive no more messages, you are just expected to *send* StatusReport. It should be an error to receive anything else.

Reply:
Changing it back to Error. The idea is that we should not receive any other message at this point. Anything we receive should be treated as error.